### ISSUE TITLE
General Questions on QIIT and Agda 

### ISSUE THREAD

**Issue by DKXXXL** [AndrasKovacs/glue#3]
Dear Dr. Andras Kovacs and Prof. Ambrus Kaposi,
    This correspondence should be more formal (i.e. in an email) but I find github thread has a cleaner way to present archive, so bear with me. I have some general questions on QIIT and its mechanization.
1. Can I have equality constructor in QIIT (in intensional MLTT) between a term and eq-coerced term, i.e. `a ≡ coe p b` where `p` itself can be from equality constructor? I heard this is possible in HIIT for Agda, but I am not sure if this is okay for QIIT because the paper I read about QIIT all works in an extensional setting. 
(Maybe the real question I am asking is if I put coe there, what is the 'signature' of the algebra?)

2. There are some cases where rewriting fails randomly. For example, I make `(p₂ [( γ ,s t)]t ) ≡  t` as a rewriting rule (along with others). At top level of the program it can reduce, but sometimes (deep in the program) it will fail

3. I see in Cockx blog post that rewriting will break Agda in some weird way (like damaging preservation and confluence, even though I am not sure what that actually means), but I do notice some weird issue happens when I put into too much rewriting rules into Agda. My question is, can I make sure all weird phenomenon go away when I open --confluence-checking ? Or I am curious how do I use rewriting safely? ( though it is also possible my Agda skill is not good enough on too many dependnet types)

Thanks,
Ende Jin

**Comment by DKXXXL** [comment]
Another question I have is still related to the signature of dependent algebra for QIIT.

I checked Prof. Kaposi's *NbE in TT* work, where the signature of dependent model seems to be defined [here](https://bitbucket.org/akaposi/tt-in-tt/src/master/TT/ElimOld.agda). I see in Agda formalization there requires an extra coercion appearing in the signature --- for example, in the syntax specification of [function beta rule](https://bitbucket.org/akaposi/tt-in-tt/src/0f083bfb58fc66aed35271b00a274feeda9fb627/TT/Func.agda#lines-31) we have a direct `app lam t = t` but its [correspondent model requires an extra coercion](https://bitbucket.org/akaposi/tt-in-tt/src/0f083bfb58fc66aed35271b00a274feeda9fb627/TT/ElimOld.agda#lines-219). 

My question is, will there be a problem if I use JMeq instead in the dependent model level? What's more, how in general can I design the signature of the dependent model of a given signature of the syntax (I am grateful for any pointer to the reference :) )

My current understanding is this mechanization is to *take Agda's as some intensional type theoretical checker for an QIIT argument in extensional type theory*. But I wonder if I can have a better and more based understanding.

I also checked Dr. Kovács work on Shallow Embedding. That is fascinating on avoiding rewriting but requires another proof of injectiveness of the standard model, a bit awkward for my current situation.


Thanks!

**Comment by ice1000** [comment]
> 3\. I see in Cockx blog post that rewriting will break Agda in some weird way (like damaging preservation and confluence, even though I am not sure what that actually means), but I do notice some weird issue happens when I put into too much rewriting rules into Agda. My question is, can I make sure all weird phenomenon go away when I open --confluence-checking ? Or I am curious how do I use rewriting safely? ( though it is also possible my Agda skill is not good enough on too many dependnet types)

I think you could ask @jespercockx.

**Comment by AndrasKovacs** [comment]
> Can I have equality constructor in QIIT (in intensional MLTT) between a term and eq-coerced term, i.e. a ≡ coe p b where p itself can be from equality constructor? 

Yes. You can replace the extensional equality with an equality supporting UIP and transport, and it would work fine. The reason for having equality reflection by default in the papers is for simplicity, and also because we need it in term algebra constructions.

> There are some cases where rewriting fails randomly.

The rewriting implementation in Agda is not flexible enough to cover all the beta-rules in a QIIT syntax for a dependent language. The details are a bit technical, the issue is in the left-hand-side matching implementation. 

> can I make sure all weird phenomenon go away when I open --confluence-checking ? Or I am curious how do I use rewriting safely?

I haven't used --confluence-checking but as far as I know it does improve on robustness significantly. It doesn't affect though the rule matching issue that I mentioned.

> My question is, will there be a problem if I use JMeq instead in the dependent model level

If you assume UIP then you can use JMeq freely, and it's a matter of taste to choose JMeq or ordinary equality. However, both are extremely tedious if you want to formalize somewhat more complicated models. Nowadays it's a bit more convenient to use a propositional equality type in `Prop` in Agda, and also postulate that we can eliminate from there to `Set`. This approximates observational/setoid type theory. It's a bit nicer because we get definitional UIP.

> how in general can I design the signature of the dependent model of a given signature of the syntax

The point is that the LHS-es and the RHS-es in equations have to be the right things, but besides that you can use transports in any way you like.

**Comment by akaposi** [comment]
> 1. Can I have equality constructor in QIIT (in intensional MLTT) between a term and eq-coerced term, i.e. `a ≡ coe p b` where `p` itself can be from equality constructor? I heard this is possible in HIIT for Agda, but I am not sure if this is okay for QIIT because the paper I read about QIIT all works in an extensional setting.
>    (Maybe the real question I am asking is if I put coe there, what is the 'signature' of the algebra?)

Yes, if you internalise the theory of signatures in an intensional metalanguage, then that coe will be the coe of the metalanguage, and it is fine. The metalanguage has to support QIITs though, e.g. cubical type theory or setoid type theory (we are working on implementing it in Haskell, see https://github.com/AndrasKovacs/sett ).

> 2. There are some cases where rewriting fails randomly. For example, I make `(p₂ [( γ ,s t)]t ) ≡  t` as a rewriting rule (along with others). At top level of the program it can reduce, but sometimes (deep in the program) it will fail
> 3. I see in Cockx blog post that rewriting will break Agda in some weird way (like damaging preservation and confluence, even though I am not sure what that actually means), but I do notice some weird issue happens when I put into too much rewriting rules into Agda. My question is, can I make sure all weird phenomenon go away when I open --confluence-checking ? Or I am curious how do I use rewriting safely? ( though it is also possible my Agda skill is not good enough on too many dependnet types)

Yes, there are issues with the implementation of rewrite rules. If you want to formalise arguments about models of type theory, I find that the shallow embedding trick is more reliable than rewrite rules (but has the downside that only works with something that has been already implemented in Agda itself, and error messages are not very helpful). The proper way to do this is to start with a QIIT-syntax (without rewrite rules), then prove normalisation (which is quite a lot of work indeed), then you can just work with normal forms. A hybrid version is where you only normalise parts of the syntax (e.g. substitution-normal forms, so obtain implicit substitutions).

> My question is, will there be a problem if I use JMeq instead in the dependent model level? What's more, how in general can I design the signature of the dependent model of a given signature of the syntax (I am grateful for any pointer to the reference :) 

Are you trying to do an Agda formalisation? JMEq is fine, but I found that currently the best thing to do is use Prop-valued equality in Agda with postulated strong transport (and rewrite rule for computation on refl) add QIITs with postulated constructors and eliminators and use rewrite rules only for the beta rules of the eliminators. See https://bitbucket.org/akaposi/typesystems/src/master/src for formalisation of simply typed lambda calculus using this approach.

**Comment by ice1000** [comment]
> The rewriting implementation in Agda is not flexible enough to cover all the beta-rules in a QIIT syntax for a dependent language. The details are a bit technical, the issue is in the left-hand-side matching implementation.

Care to throw some random counterexamples/papers/links?

**Comment by ice1000** [comment]
But speaking of the Agda `Prop`, is it okay metatheoretically to have strong transport? I think the weak `Prop` in Coq has that. Lean's Prop is even stronger, but that relies on a subsingleton test which I think is fragile.

**Comment by AndrasKovacs** [comment]
The strong transport is specifically postulated for the identity type, and we don't have any general mechanism to eliminate from Prop to Set. Metatheoretically, this is validated by a setoid model or a model in ETT. The only ways to eliminate from Prop to Set are from identity and the empty Prop.

**Comment by ice1000** [comment]
> The strong transport is specifically postulated for the identity type, and we don't have any general mechanism to eliminate from Prop to Set.

Yes. I'm talking about exactly that -- not general mechanism, but specifically about identity (strong transport or rewrite). I expect an observational equality-ish equality type which is built-in and has special support in core so that you can rewrite or strong transport. I'm wondering if that makes semantic sense, because I can be wrong assuming observational equality being compatible with setoidtt (I only have basic intuition, I haven't read the papers).

**Comment by AndrasKovacs** [comment]
> Care to throw some random counterexamples/papers/links?

```
{-# OPTIONS --rewriting #-}

open import Data.Bool
open import Relation.Binary.PropositionalEquality

{-# BUILTIN REWRITE _≡_ #-}

postulate
  f    : Bool → Bool → Bool
  rule : ∀ b → f (not b) b ≡ false
{-# REWRITE rule #-}

fail : f true false ≡ false
fail = {!refl!}
```

Here, Agda fails to match `true` to `not b`. A better matching implementation would postpone such matching failures until all arguments are processed.

**Comment by jespercockx** [comment]
> > 3. I see in Cockx blog post that rewriting will break Agda in some weird way (like damaging preservation and confluence, even though I am not sure what that actually means), but I do notice some weird issue happens when I put into too much rewriting rules into Agda. My question is, can I make sure all weird phenomenon go away when I open --confluence-checking ? Or I am curious how do I use rewriting safely? ( though it is also possible my Agda skill is not good enough on too many dependnet types)
> 
> I think you could ask @jespercockx.

You're correct that rewrite rules can break confluence, and through that also break subject reduction, if you are not careful. The `--confluence-check` flag should detect all of the bad cases though, so as long as that doesn't complain you should be fine. The other problems that might arise from rewrite rules (and that the confluence checker does not protect against) are (1) risk of non-terminating type checking (if you add a non-terminating rule such as `x + y --> y + x`) and incompatibility with certain axioms (if you rewrite by equalities constructed with univalence).

**Comment by jespercockx** [comment]
> > Care to throw some random counterexamples/papers/links?
> 
> ```
> {-# OPTIONS --rewriting #-}
> 
> open import Data.Bool
> open import Relation.Binary.PropositionalEquality
> 
> {-# BUILTIN REWRITE _≡_ #-}
> 
> postulate
>   f    : Bool → Bool → Bool
>   rule : ∀ b → f (not b) b ≡ false
> {-# REWRITE rule #-}
> 
> fail : f true false ≡ false
> fail = {!refl!}
> ```
> 
> Here, Agda fails to match `true` to `not b`. A better matching implementation would postpone such matching failures until all arguments are processed.

This is not accepted by the confluence checker though, and you can fix it by adding extra rules for `f true false` and `f false true` to resolve the critical pairs.

**Comment by AndrasKovacs** [comment]
It's been a while since I looked at this, but I think that for embedded QIIT beta rules the issue was that we needed an unbounded number of specialized rules.

I'll try to look at this again, now with confluence checking. I think that there should be confluent examples too which did not compute sufficiently.

**Comment by DKXXXL** [comment]
Thanks for the reply! These are all very useful information!
I also have some confusion about, the meta-theoretically reason, why I can have a syntax (without any coercion) of `lam app t = t` but a corresponding model has an extra coercion 
(for example, the syntax specification of [function beta rule](https://bitbucket.org/akaposi/tt-in-tt/src/0f083bfb58fc66aed35271b00a274feeda9fb627/TT/Func.agda#lines-31) but [the displayed algebra requires an extra coercion](https://bitbucket.org/akaposi/tt-in-tt/src/0f083bfb58fc66aed35271b00a274feeda9fb627/TT/ElimOld.agda#lines-219).

It is like,  I cannot intuitively imagine what this section of this displayed algebra (or eliminator?) looks like. 

What's more, if in two models, the coercion are using two different 'path‘ (two different proof of equality, I am not HoTT guy though barely speaking of the terminology 'path'), are they both models of the syntax (without any coercion)?

Any reference pointer is welcomed! :)

Thanks!

**Comment by ice1000** [comment]
I believe that in cubical, having Kan operations in path constructors is fine, it just makes things tedious (but not break things). Consider the definition of torus from the hott book -- it is very tedious to work with in cubical.

**Comment by AndrasKovacs** [comment]
> why I can have a syntax (without any coercion) of lam app t = t but a corresponding model has an extra coercion

Every component of a displayed model is "over" the correspond component of the base model. An equation being over a base equation means that it's a dependent equality. Without the transport over the base a displayed equation can't even be well-typed. The transport can be only omitted if the base equation holds definitionally.

**Comment by DKXXXL** [comment]
> > why I can have a syntax (without any coercion) of lam app t = t but a corresponding model has an extra coercion
> 
> Every component of a displayed model is "over" the correspond component of the base model. An equation being over a base equation means that it's a dependent equality. Without the transport over the base a displayed equation can't even be well-typed. The transport can be only omitted if the base equation holds definitionally.

Yes. To have that proposition well-typed I definitely need equality coercion. I am more of asking (any literature reference for) why meta-theoretically, there is a map from the syntax to the displayed model (i.e. the existence of section). 

Surprisingly in extensional setting, I don't find that counter-intuitive at all. For me, intuitively, this is a mapping of 2 parts: (1) first mapping syntax (data) (without equation) to an algebra (without equation), and  (2) those mapping from equations to equations is just to make sure the mapping function itself makes sense otherwise the consistency of the whole system will collapse.

I think "Constructing QIIT" paper was about justifying the existence of section in extensional setting.

However, once coercion comes in, in an intensional setting like Agda, my brain fried.

**Comment by DKXXXL** [comment]
> > 1. Can I have equality constructor in QIIT (in intensional MLTT) between a term and eq-coerced term, i.e. `a ≡ coe p b` where `p` itself can be from equality constructor? I heard this is possible in HIIT for Agda, but I am not sure if this is okay for QIIT because the paper I read about QIIT all works in an extensional setting.
> >    (Maybe the real question I am asking is if I put coe there, what is the 'signature' of the algebra?)
> 
> Yes, if you internalise the theory of signatures in an intensional metalanguage, then that coe will be the coe of the metalanguage, and it is fine. The metalanguage has to support QIITs though, e.g. cubical type theory or setoid type theory (we are working on implementing it in Haskell, see https://github.com/AndrasKovacs/sett ).
> 
> > 2. There are some cases where rewriting fails randomly. For example, I make `(p₂ [( γ ,s t)]t ) ≡  t` as a rewriting rule (along with others). At top level of the program it can reduce, but sometimes (deep in the program) it will fail
> > 3. I see in Cockx blog post that rewriting will break Agda in some weird way (like damaging preservation and confluence, even though I am not sure what that actually means), but I do notice some weird issue happens when I put into too much rewriting rules into Agda. My question is, can I make sure all weird phenomenon go away when I open --confluence-checking ? Or I am curious how do I use rewriting safely? ( though it is also possible my Agda skill is not good enough on too many dependnet types)
> 
> Yes, there are issues with the implementation of rewrite rules. If you want to formalise arguments about models of type theory, I find that the shallow embedding trick is more reliable than rewrite rules (but has the downside that only works with something that has been already implemented in Agda itself, and error messages are not very helpful). The proper way to do this is to start with a QIIT-syntax (without rewrite rules), then prove normalisation (which is quite a lot of work indeed), then you can just work with normal forms. A hybrid version is where you only normalise parts of the syntax (e.g. substitution-normal forms, so obtain implicit substitutions).
> 
> > My question is, will there be a problem if I use JMeq instead in the dependent model level? What's more, how in general can I design the signature of the dependent model of a given signature of the syntax (I am grateful for any pointer to the reference :)
> 
> Are you trying to do an Agda formalisation? JMEq is fine, but I found that currently the best thing to do is use Prop-valued equality in Agda with postulated strong transport (and rewrite rule for computation on refl) add QIITs with postulated constructors and eliminators and use rewrite rules only for the beta rules of the eliminators. See https://bitbucket.org/akaposi/typesystems/src/master/src for formalisation of simply typed lambda calculus using this approach.

Greatly thankful for the reference! I am currently checking this repo.
I have a quick question, in `Lib.lagda` [there are](https://bitbucket.org/akaposi/typesystems/src/master/src/Lib.lagda) two notions of equality, proof relevant `_≈_` and irrelevant one `_≡_`. They seem both used. I wonder why especially the proof-relevant one is chosen as the `BUILTIN REWRITE`? 
(Though I see there is [some legacy code](https://bitbucket.org/akaposi/typesystems/src/master/archive/tut-2021/gy01_after.agda) taking the irrelevant-one as `BUILTIN REWRITE` but it seems to be archived)

I mean, my first intuition is just using one equality (of course the proof irrelevant one) and then use that as `BUILTIN REWRITE` and all the rewrite rules will be using `_≈_`. Is this intuition problematic somehow but I failed to see?
 

EDIT: What's more, actually I also think if I want to reduce the number of `coe`/`transpose` in syntax signature by using rewrite, I will have to use `BUILTIN REWRITE` on the irrelevant one `_≡_`. 

Finally, correct me if I am wrong, the syntax about the equality proposed by [Setoid Type Theory](https://hal.inria.fr/hal-02281225) paper (section 5) , it is actually an proof-irrelevant JMeq right?

**Comment by akaposi** [comment]
> I mean, my first intuition is just using one equality (of course the proof irrelevant one) and then use that as `BUILTIN REWRITE` and all the rewrite rules will be using `_≈_`. Is this intuition problematic somehow but I failed to see?

Yes, `_≈_` is only there to specify some rewrite rules. But I found that it is best to keep rewriting absolutely minimal: so I only rewrite with transportrefl and the computation rules of QIITs.

> EDIT: What's more, actually I also think if I want to reduce the number of `coe`/`transpose` in syntax signature by using rewrite, I will have to use `BUILTIN REWRITE` on the irrelevant one `_≡_`.

In the syntax (in a QIIT) I don't want to use rewriting because the setoid/cubical set model only justifies these equalities weakly and not strictly. You can strictify syntaxes though after defining them without rewrite rules: you define (partial) normalisation and prove that these normal forms have the same elimination principle as the original syntax. Then you can just use the normal syntax. For example, [here](https://bitbucket.org/akaposi/qiirt/src/master/QIIRT-postulated.agda) I replace explicit substitution by a recursively defined one. The notion of dependent/displayed model over this new syntax has fewer transports, thus the canonicity proof is slightly easier: compare [this](https://bitbucket.org/akaposi/qiirt/src/master/QIIRT-postulated/Canon.agda) and [this](https://bitbucket.org/akaposi/qiirt/src/master/QIIT-postulated/Canon.agda).

> Finally, correct me if I am wrong, the syntax about the equality proposed by [Setoid Type Theory](https://hal.inria.fr/hal-02281225) paper (section 5) , it is actually an proof-irrelevant JMeq right?

It is not John Major, it is a different kind of heterogeneous equality where you only have Id type for provably equal types, not arbitrary two types.
